From 665d6c6e892e15e9d642a03f16b2425cb1c4e4fe Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Mon, 30 Jun 2025 13:55:14 +0800
Subject: [PATCH 104/128] wifi: mac80211: do not update channel if csa ie
 exists

Station may receive beacon ie with old channel after channel switch.
Do not update channel when CSA ie exists to prevent station
switch back to old channel and leads to disconnection.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 net/mac80211/mlme.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/net/mac80211/mlme.c b/net/mac80211/mlme.c
index 2f5ea56f..939cb386 100644
--- a/net/mac80211/mlme.c
+++ b/net/mac80211/mlme.c
@@ -7775,7 +7775,8 @@ static void ieee80211_rx_mgmt_beacon(struct ieee80211_link_data *link,
 
 	changed |= ieee80211_recalc_twt_req(sdata, sband, link, link_sta, elems);
 
-	if (ieee80211_config_bw(link, elems, true, &changed,
+	if (!elems->ch_switch_ie &&
+	    ieee80211_config_bw(link, elems, true, &changed,
 				IEEE80211_STYPE_BEACON)) {
 		ieee80211_set_disassoc(sdata, IEEE80211_STYPE_DEAUTH,
 				       WLAN_REASON_DEAUTH_LEAVING,
-- 
2.45.2

